### PR TITLE
[FLINK-14711][table] add alter and show function ddl

### PR DESCRIPTION
**Pull request by HuangZhenQiu**
## Brief change log
  -  Add alter function and show functions ddl in SQL parser
  -  Add unit test in FlinkSqlParserImplTest 

## Verifying this change

The alter and show function ddl syntax is verified in unit tests of FlinkSqlParserImplTest.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by HuangZhenQiu** [comment]
@zjuwangg  @bowenli86 @xuefuz 
Please review this PR at your most convenient time. The PR only add new functions in Parser. As the https://github.com/apache/flink/pull/9689/files has language-related definition. I will bridge alter and show functions with runtime after 9689 is merged. Thanks.

File: flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl [hunk 1]
```diff
@@ -177,6 +177,56 @@ SqlDescribeDatabase SqlDescribeDatabase() :
...
+    boolean isSystemFunction = false;
+}
+{
+    <ALTER>
+    <TEMPORARY>
+    (
+        <SYSTEM>   { isSystemFunction = true; }
+    |
+        {isSystemFunction = false; }
+    )
```
Comment (zjuwangg) [inline flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl]: Better be replaced with
`
[ <SYSTEM> { isSystemFunction = true}]
`
?

File: flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl [hunk 2]
```diff
@@ -177,6 +177,56 @@ SqlDescribeDatabase SqlDescribeDatabase() :
...
+    <FUNCTION> { startPos = getPos(); }
+    (
+        <IF> <EXISTS> { ifExists = true; }
+    |
+        { ifExists = false; }
+    )
```
Comment (zjuwangg) [inline flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl]: ditto.

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterDatabase.java [hunk 3]
```diff
@@ -41,7 +41,6 @@
 	public static final SqlSpecialOperator OPERATOR = new SqlSpecialOperator("ALTER DATABASE", SqlKind.OTHER);
 
 	private final SqlIdentifier databaseName;
-
```
Comment (zjuwangg) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterDatabase.java]: Unrelated modification?

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterFunction.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+	@Nonnull
+	@Override
+	public List<SqlNode> getOperandList() {
+		return ImmutableNullableList.of(functionName, functionClassName, functionLanguage);
+	}
+
+	public String[] getFullFunctionName() {
+		return functionName.names.toArray(new String[0]);
+	}
+}
```
Comment (zjuwangg) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterFunction.java]: Should this class provide a getter for `language`?

**Comment by zjuwangg** [comment]
Thanks @HuangZhenQiu for your work on this. 
I left some comments from my side.

File: flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl [hunk 5]
```diff
@@ -177,6 +177,56 @@ SqlDescribeDatabase SqlDescribeDatabase() :
...
+    functionName = CompoundIdentifier()
+    [ <AS> <QUOTED_STRING> {
+        String p = SqlParserUtil.parseString(token.image);
+        functionClassName = SqlLiteral.createCharString(p, getPos());
+    }]
+    [ <LANGUAGE> <QUOTED_STRING> {
```
Comment (xuefuz) [inline flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl]: Not sure if what's our decision on language types, but have a quoted string for language seems suggestion user can put anything. One alternative is define constant such as SQL, PYTHON, JAVA, etc.

File: flink-table/flink-sql-parser/src/main/codegen/data/Parser.tdd [hunk 6]
```diff
@@ -342,7 +346,6 @@
     "SUBCLASS_ORIGIN"
     "SUBSTITUTE"
     "TABLE_NAME"
-    "TEMPORARY"
```
Comment (xuefuz) [inline flink-table/flink-sql-parser/src/main/codegen/data/Parser.tdd]: What's the reason behind this change?

File: flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl [hunk 7]
```diff
@@ -177,6 +177,56 @@ SqlDescribeDatabase SqlDescribeDatabase() :
...
+    SqlIdentifier functionName = null;
+    SqlCharStringLiteral functionClassName = null;
+    SqlCharStringLiteral functionLanguage = null;
+    SqlParserPos startPos;
+    boolean ifExists = false;
+    boolean isSystemFunction = false;
+}
+{
+    <ALTER>
+    <TEMPORARY>
```
Comment (xuefuz) [inline flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl]: is <TEMPORARY> optional for alter function? My understanding is that both <TEMPORARY> and <SYSTEM> should be optional. The grammar here suggest <TEMPORARY> is required.

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterFunction.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Alter Function Sql Call.
+ */
+public class SqlAlterFunction extends SqlCall {
+	public static final SqlSpecialOperator OPERATOR = new SqlSpecialOperator("ALTER FUNCTION", SqlKind.OTHER);
+	private final SqlIdentifier functionName;
```
Comment (xuefuz) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterFunction.java]: Nit: an empty line before this would make the code more readable.

File: flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl [hunk 9]
```diff
@@ -177,6 +177,56 @@ SqlDescribeDatabase SqlDescribeDatabase() :
...
+        functionLanguage = SqlLiteral.createCharString(lang, getPos());
+    }]
+    {
+        return new SqlAlterFunction(startPos.plus(getPos()), functionName, functionClassName, functionLanguage, ifExists, isSystemFunction);
+    }
+}
+
+SqlShowFunctions SqlShowFunctions() :
+{
+    SqlIdentifier functionScope = null;
```
Comment (xuefuz) [inline flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl]: Calling this functionScope might not be actuate. It should be just "database".

**Comment by HuangZhenQiu** [comment]
@zjuwangg @xuefuz 
Thanks for your comments. I updated the PR accordingly. Please review again at your most convenient time.

File: flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl [hunk 10]
```diff
@@ -177,6 +177,65 @@ SqlDescribeDatabase SqlDescribeDatabase() :
...
 }
 
+SqlAlterFunction SqlAlterFunction() :
+{
+    SqlIdentifier functionName = null;
+    SqlCharStringLiteral functionClassName = null;
+    String functionLanguage = null;
+    SqlParserPos startPos;
+    boolean ifExists = false;
+    boolean hasTemporary = false;
```
Comment (bowenli86) [inline flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl]: isTemporary?

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterFunction.java [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+	public SqlAlterFunction(
+		SqlParserPos pos,
+		SqlIdentifier functionName,
+		SqlCharStringLiteral functionClassName,
+		String functionLanguage,
+		boolean ifExists,
+		boolean hasTemporary,
+		boolean isSystemFunction) {
+		super(pos);
```
Comment (bowenli86) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterFunction.java]: minor: one more tab for this section

File: flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl [hunk 12]
```diff
@@ -177,6 +177,65 @@ SqlDescribeDatabase SqlDescribeDatabase() :
...
+
+    <AS> <QUOTED_STRING> {
+        String p = SqlParserUtil.parseString(token.image);
+        functionClassName = SqlLiteral.createCharString(p, getPos());
+    }
+
+    [<LANGUAGE>
+        (   <JAVA>  { functionLanguage = "JAVA"; }
+        |
+            <SCALA> { functionLanguage = "SCALA"; }
```
Comment (xuefuz) [inline flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl]: Did we say "SCALA = JAVA"?

File: flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl [hunk 13]
```diff
@@ -177,6 +177,65 @@ SqlDescribeDatabase SqlDescribeDatabase() :
...
+    boolean isSystemFunction = false;
+}
+{
+    <ALTER>
+
+    [ <TEMPORARY> { hasTemporary = true; }
+        (
+            <SYSTEM>   { isSystemFunction = true; }
+        |
+            {isSystemFunction = false; }
```
Comment (xuefuz) [inline flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl]: I think this can be skipped if we initialize isSystemFunction to false.

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowFunctions.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+		SqlWriter writer,
+		int leftPrec,
+		int rightPrec) {
+		writer.keyword("SHOW FUNCTIONS");
+		if (databaseName != null) {
+			databaseName.unparse(writer, leftPrec, rightPrec);
+		}
+	}
+
+	public String[] fullDatabaseName() {
```
Comment (xuefuz) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowFunctions.java]: Nit: we might want to call it getXXX() instead. I'm not sure of "full", as the database might be just "db" or "cat.db".

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterFunction.java [hunk 15]
```diff
@@ -52,21 +52,21 @@
...
 	public SqlAlterFunction(
-		SqlParserPos pos,
-		SqlIdentifier functionName,
-		SqlCharStringLiteral functionClassName,
-		String functionLanguage,
-		boolean ifExists,
-		boolean hasTemporary,
-		boolean isSystemFunction) {
+			SqlParserPos pos,
+			SqlIdentifier functionName,
```
Comment (bowenli86) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterFunction.java]: minor: rename to `functionIdentifier`?

File: flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl [hunk 16]
```diff
@@ -177,6 +177,61 @@ SqlDescribeDatabase SqlDescribeDatabase() :
 
 }
 
+SqlAlterFunction SqlAlterFunction() :
+{
+    SqlIdentifier functionName = null;
```
Comment (bowenli86) [inline flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl]: rename to `functionIdentifier`?

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterFunction.java [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+	@Override
+	public List<SqlNode> getOperandList() {
+		return ImmutableNullableList.of(functionName, functionClassName);
+	}
+
+	public String getLanguage() {
+		return functionLanguage;
+	}
+
+	public String[] getFunctionName() {
```
Comment (bowenli86) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/ddl/SqlAlterFunction.java]: rename to `getFunctionIdentifier()`?

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowFunctions.java [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+			SqlWriter writer,
+			int leftPrec,
+			int rightPrec) {
+		writer.keyword("SHOW FUNCTIONS");
+		if (databaseName != null) {
+			databaseName.unparse(writer, leftPrec, rightPrec);
+		}
+	}
+
+	public String[] getDatabaseName() {
```
Comment (bowenli86) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowFunctions.java]: rename to `getDatabasePath()`?

**Comment by HuangZhenQiu** [comment]
@bowenli86 @xuefuz Thanks for the kindly review. 
I just updated the naming of functions accordingly.

**Comment by bowenli86** [comment]
LGTM. Just a minor issue: please use the ticket title as the first commit message, e.g. "[FLINK-14711][table] add alter and show function ddl"
